File: packages/server/cmd/renewlet/import_export.go
package main

// import_export.go 实现 Renewlet/Wallos 导入预览与执行。
//
// 架构位置：
//   - 前端先在浏览器本地解析文件，再把标准 importPayload 交给这里做用户隔离、冲突预览和写库。
//   - extra.import 是跨 Go/PocketBase、Cloudflare Worker 与前端 shared schema 的幂等键事实来源。
//   - apply 会重新 preview 并在事务内写 subscriptions/settings/custom_configs，避免 UI 预览被篡改后直接落库。
//
// 注意： 预览上限服务于冲突分析，执行上限服务于真实写库成本；两者不要合并成一个魔法数字。

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

const maxImportJSONBodyBytes int64 = 50 << 20
const maxImportPreviewSubscriptions = 5000
const maxImportApplySubscriptions = 200
const importExistingPageSize = 500
const importWarningLowConfidenceKey = "IMPORT_WARNING_LOW_CONFIDENCE_KEY"
const importWarningLowConfidenceNameMatched = "IMPORT_WARNING_LOW_CONFIDENCE_NAME_MATCHED"

type importPreviewRequest struct {
	Payload      importPayload `json:"payload"`
	ConflictMode string        `json:"conflictMode"`
	SkipIndexes  []int         `json:"skipIndexes,omitempty"`
}

type importApplyRequest struct {
	Payload      importPayload `json:"payload"`
	ConflictMode string        `json:"conflictMode"`
	SkipIndexes  []int         `json:"skipIndexes,omitempty"`
}

type importPayload struct {
	Source        string               `json:"source"`
	Subscriptions []importSubscription `json:"subscriptions"`
	Settings      json.RawMessage      `json:"settings,omitempty"`
	CustomConfig  *customConfigPayload `json:"customConfig,omitempty"`
}

type importSubscription struct {
	Name                         string                 `json:"name"`
	Logo                         *string                `json:"logo,omitempty"`
	Price                        float64                `json:"price"`
	Currency                     string                 `json:"currency"`
	BillingCycle                 string                 `json:"billingCycle"`
	CustomDays                   *int                   `json:"customDays,omitempty"`
	Category                     string                 `json:"category"`
	Status                       string                 `json:"status"`
	Pinned                       bool                   `json:"pinned"`
	PaymentMethod                *string                `json:"paymentMethod,omitempty"`
	StartDate                    string                 `json:"startDate"`
	NextBillingDate              string                 `json:"nextBillingDate"`
	AutoCalculateNextBillingDate bool                   `json:"autoCalculateNextBillingDate"`
	TrialEndDate                 *string                `json:"trialEndDate,omitempty"`
	Website                      *string                `json:"website,omitempty"`
	Notes                        *string                `json:"notes,omitempty"`
	Tags                         []string               `json:"tags,omitempty"`
	ReminderDays                 int                    `json:"reminderDays"`
	RepeatReminderEnabled        bool                   `json:"repeatReminderEnabled"`
	RepeatReminderInterval       string                 `json:"repeatReminderInterval"`
	RepeatReminderWindow         string                 `json:"repeatReminderWindow"`
	Extra                        map[string]interface{} `json:"extra"`
}

type importPreviewResponse struct {
	Summary              importSummary       `json:"summary"`
	Items                []importPreviewItem `json:"items"`
	IncludesSettings     bool                `json:"includesSettings"`
	IncludesCustomConfig bool                `json:"includesCustomConfig"`
}

type importApplyResponse struct {
	OK bool `json:"ok"`
	importPreviewResponse
}

type importPreviewItem struct {
	Index      int      `json:"index"`
	Name       string   `json:"name"`
	Source     string   `json:"source"`
	SourceID   string   `json:"sourceId"`
	ExistingID string   `json:"existingId,omitempty"`
	Action     string   `json:"action"`
	Warnings   []string `json:"warnings"`
	Errors     []string `json:"errors"`
}

type importSummary struct {
	Total    int `json:"total"`
	Creates  int `json:"creates"`
	Replaces int `json:"replaces"`
	Skips    int `json:"skips"`
	Errors   int `json:"errors"`
	Warnings int `json:"warnings"`
}

type importKey struct {
	Source     string
	SourceID   string
	Confidence string
}

type importExistingMatches struct {
	ByKey                   map[string]*core.Record
	LowConfidenceByName     map[string]*core.Record
	LowConfidenceDuplicates map[string]bool
}

func (r *importPreviewRequest) Validate(locale appLocale) error {
	return validateImportPayload(r.Payload, r.ConflictMode, r.SkipIndexes, maxImportPreviewSubscriptions, locale)
}

func (r *importApplyRequest) Validate(locale appLocale) error {
	return validateImportPayload(r.Payload, r.ConflictMode, r.SkipIndexes, maxImportApplySubscriptions, locale)
}

func handleImportPreview(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	// 导入 payload 不包含二进制资产，但 5000 条订阅和 extra 元数据会超过普通 API 的 1MiB 上限。
	body, err := decodeStrictJSONWithLimit[importPreviewRequest](e.Request, locale, maxImportJSONBodyBytes)
	if err != nil {
		return e.BadRequestError(validationErrorMessage(locale, "common.invalidRequestBody", err), err)
	}
	response, err := previewImportPayload(app, e.Auth, body.Payload, body.ConflictMode, body.SkipIndexes)
	if err != nil {
		return e.BadRequestError(serverText(locale, "import.invalid"), err)
	}
	return e.JSON(http.StatusOK, response)
}

func handleImportApply(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	// apply 会重新预览再进事务，防止调用方篡改 preview 结果后直接写库。
	body, err := decodeStrictJSONWithLimit[importApplyRequest](e.Request, locale, maxImportJSONBodyBytes)
	if err != nil {
		return e.BadRequestError(validationErrorMessage(locale, "common.invalidRequestBody", err), err)
	}
	preview, err := previewImportPayload(app, e.Auth, body.Payload, body.ConflictMode, body.SkipIndexes)
	if err != nil {
		return e.BadRequestError(serverText(locale, "import.invalid"), err)
	}
	if preview.Summary.Errors > 0 {
		return e.BadRequestError(serverText(locale, "import.payloadContainsErrors"), preview)
	}
	if err := applyImportPayload(app, e.Auth, body.Payload, body.ConflictMode, body.SkipIndexes); err != nil {
		return e.BadRequestError(serverText(locale, "import.applyFailed"), err)
	}
	return e.JSON(http.StatusOK, importApplyResponse{OK: true, importPreviewResponse: preview})
}

func validateImportPayload(payload importPayload, conflictMode string, skipIndexes []int, maxSubscriptions int, _ appLocale) error {
	if conflictMode != "replace" && conflictMode != "skip" {
		return errors.New("IMPORT_CONFLICT_MODE_INVALID")
	}
	if payload.Source != "renewlet" && payload.Source != "wallos" {
		return errors.New("IMPORT_SOURCE_INVALID")
	}
	if len(payload.Subscriptions) > maxSubscriptions {
		return errors.New("IMPORT_TOO_MANY_SUBSCRIPTIONS")
	}
	if _, err := importSkippedIndexSet(skipIndexes, len(payload.Subscriptions)); err != nil {
		return err
	}
	if rawJSONIsNull(payload.Settings) {
		return errors.New("IMPORT_SETTINGS_INVALID")
	}
	for i := range payload.Subscriptions {
		key, err := importKeyFromExtra(payload.Subscriptions[i].Extra)
		if err != nil {
			return fmt.Errorf("subscription %d: %w", i+1, err)
		}
		if key.Source != payload.Source {
			return fmt.Errorf("subscription %d: IMPORT_SOURCE_MISMATCH", i+1)
		}
	}
	if payload.CustomConfig != nil {
		if err := normalizeCustomConfigPayload(payload.CustomConfig); err != nil {
			return err
		}
	}
	return nil
}

func previewImportPayload(app core.App, user *core.Record, payload importPayload, conflictMode string, skipIndexes []int) (importPreviewResponse, error) {
	rows, err := listImportExistingSubscriptions(app, user.Id)
	if err != nil {
		return importPreviewResponse{}, err
	}
	skippedIndexes, err := importSkippedIndexSet(skipIndexes, len(payload.Subscriptions))
	if err != nil {
		return importPreviewResponse{}, err
	}
	existingMatches := existingSubscriptionMatches(rows)
	items := make([]importPreviewItem, 0, len(payload.Subscriptions))
	seenPayloadKeys := map[string]bool{}
	for index := range payload.Subscriptions {
		subscription := payload.Subscriptions[index]
		key, keyErr := importKeyFromExtra(subscription.Extra)
		item := importPreviewItem{
			Index:    index,
			Name:     strings.TrimSpace(subscription.Name),
			Warnings: []string{},
			Errors:   []string{},
		}
		if keyErr != nil {
			item.Action = "error"
			item.Errors = append(item.Errors, keyErr.Error())
			items = append(items, item)
			continue
		}
		item.Source = key.Source
		item.SourceID = key.SourceID
		if key.Confidence == "low" {
			item.Warnings = append(item.Warnings, importWarningLowConfidenceKey)
		}
		if skippedIndexes[index] {
			item.Action = "skip"
			items = append(items, item)
			continue
		}
		keyString := importKeyString(key)
		if seenPayloadKeys[keyString] {
			// 单个导入文件里的重复幂等键必须失败；否则 replace 会把两条来源记录写到同一订阅。
			item.Action = "error"
			item.Errors = append(item.Errors, "IMPORT_SOURCE_ID_DUPLICATE")
			items = append(items, item)
			continue
		}
		seenPayloadKeys[keyString] = true
		if err := validateImportSubscription(app, user, subscription); err != nil {
			item.Action = "error"
			item.Errors = append(item.Errors, err.Error())
			items = append(items, item)
			continue
		}
		if existing, fallback := existingMatches.Resolve(key, subscription); existing != nil {
			item.ExistingID = existing.Id
			if fallback {
				// Wallos display:* 只能按名称低置信桥接，给 warning 让用户确认，不把它伪装成精确命中。
				item.Warnings = append(item.Warnings, importWarningLowConfidenceNameMatched)
			}
			if conflictMode == "replace" {
				item.Action = "replace"
			} else {
				item.Action = "skip"
			}
		} else {
			item.Action = "create"
		}
		items = append(items, item)
	}
	return importPreviewResponse{
		Summary:              summarizeImportItems(items),
		Items:                items,
		IncludesSettings:     len(strings.TrimSpace(string(payload.Settings))) > 0,
		IncludesCustomConfig: payload.CustomConfig != nil,
	}, nil
}

func applyImportPayload(app core.App, user *core.Record, payload importPayload, conflictMode string, skipIndexes []int) error {
	// 导入写入包在 PocketBase 事务内完成；任意订阅、settings 或 custom config 失败都不能留下半套迁移数据。
	return app.RunInTransaction(func(txApp core.App) error {
		rows, err := listImportExistingSubscriptions(txApp, user.Id)
		if err != nil {
			return err
		}
		collection, err := txApp.FindCollectionByNameOrId("subscriptions")
		if err != nil {
			return err
		}
		skippedIndexes, err := importSkippedIndexSet(skipIndexes, len(payload.Subscriptions))
		if err != nil {
			return err
		}
		existingMatches := existingSubscriptionMatches(rows)
		for index, subscription := range payload.Subscriptions {
			if skippedIndexes[index] {
				continue
			}
			key, err := importKeyFromExtra(subscription.Extra)
			if err != nil {
				return err
			}
			existing, _ := existingMatches.Resolve(key, subscription)
			if existing != nil && conflictMode == "skip" {
				continue
			}
			record := existing
			if record == nil {
				record = core.NewRecord(collection)
			}
			setImportSubscriptionRecord(record, user.Id, subscription)
			if err := txApp.Save(record); err != nil {
				return err
			}
		}
		if err := applyImportedSettings(txApp, user, payload.Settings); err != nil {
			return err
		}
		if err := applyImportedCustomConfig(txApp, user, payload.CustomConfig); err != nil {
			return err
		}
		return nil
	})
}

func listImportExistingSubscriptions(app core.App, userID string) ([]*core.Record, error) {
	rows := []*core.Record{}
	for offset := 0; ; offset += importExistingPageSize {
		page, err := app.FindRecordsByFilter("subscriptions", "user = {:user}", "-created", importExistingPageSize, offset, dbx.Params{"user": userID})
		if err != nil {
			return nil, err
		}
		rows = append(rows, page...)
		if len(page) < importExistingPageSize {
			return rows, nil
		}
	}
}

func validateImportSubscription(app core.App, user *core.Record, subscription importSubscription) error {
	collection, err := app.FindCollectionByNameOrId("subscriptions")
	if err != nil {
		return err
	}
	if !isValidBillingCycle(subscription.BillingCycle) {
		return errors.New("BILLING_CYCLE_INVALID")
	}
	if !isValidSubscriptionStatus(subscription.Status) {
		return errors.New("SUBSCRIPTION_STATUS_INVALID")
	}
	record := core.NewRecord(collection)
	setImportSubscriptionRecord(record, user.Id, subscription)
	// 预览只校验不写库；复用 hooks 的核心规范化，保证 Docker 与普通订阅写入边界一致。
	return normalizeSubscriptionRecord(record)
}

func setImportSubscriptionRecord(record *core.Record, userID string, subscription importSubscription) {
	record.Set("user", userID)
	record.Set("name", subscription.Name)
	record.Set("logo", optionalString(subscription.Logo))
	record.Set("price", subscription.Price)
	record.Set("currency", subscription.Currency)
	record.Set("billingCycle", subscription.BillingCycle)
	if subscription.CustomDays != nil {
		record.Set("customDays", *subscription.CustomDays)
	} else {
		record.Set("customDays", 0)
	}
	record.Set("category", subscription.Category)
	record.Set("status", subscription.Status)
	record.Set("pinned", subscription.Pinned)
	record.Set("paymentMethod", optionalString(subscription.PaymentMethod))
	record.Set("startDate", subscription.StartDate)
	record.Set("nextBillingDate", subscription.NextBillingDate)
	record.Set("autoCalculateNextBillingDate", subscription.AutoCalculateNextBillingDate)
	record.Set("trialEndDate", optionalString(subscription.TrialEndDate))
	record.Set("website", optionalString(subscription.Website))
	record.Set("notes", optionalString(subscription.Notes))
	record.Set("tags", subscription.Tags)
	record.Set("reminderDays", subscription.ReminderDays)
	record.Set("repeatReminderEnabled", subscription.RepeatReminderEnabled)
	record.Set("repeatReminderInterval", subscription.RepeatReminderInterval)
	record.Set("repeatReminderWindow", subscription.RepeatReminderWindow)
	// extra.import 是导入唯一同源键；只写 allowlist 字段后再整体存入 JSON，避免用户 payload 扩权。
	record.Set("extra", subscription.Extra)
}

func applyImportedSettings(app core.App, user *core.Record, raw json.RawMessage) error {
	if len(strings.TrimSpace(string(raw))) == 0 {
		return nil
	}
	current := defaultAppSettings()
	record, err := app.FindFirstRecordByFilter("settings", "user = {:user}", dbx.Params{"user": user.Id})
	if err == nil && record != nil {
		current = settingsFromRecord(record)
	} else if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return err
	}
	next, err := mergeSettings(current, raw)
	if err != nil {
		return err
	}
	if record == nil {
		collection, err := app.FindCollectionByNameOrId("settings")
		if err != nil {
			return err
		}
		record = core.NewRecord(collection)
		record.Set("user", user.Id)
	}
	record.Set("settings", next)
	return app.Save(record)
}

func applyImportedCustomConfig(app core.App, user *core.Record, config *customConfigPayload) error {
	if config == nil {
		return nil
	}
	if err := normalizeCustomConfigPayload(config); err != nil {
		return err
	}
	record, err := app.FindFirstRecordByFilter("custom_configs", "user = {:user}", dbx.Params{"user": user.Id})
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return err
	}
	if record == nil {
		collection, err := app.FindCollectionByNameOrId("custom_configs")
		if err != nil {
			return err
		}
		record = core.NewRecord(collection)
		record.Set("user", user.Id)
	}
	record.Set("config", config)
	return app.Save(record)
}

func existingSubscriptionMatches(rows []*core.Record) importExistingMatches {
	result := importExistingMatches{
		ByKey:                   map[string]*core.Record{},
		LowConfidenceByName:     map[string]*core.Record{},
		LowConfidenceDuplicates: map[string]bool{},
	}
	for _, row := range rows {
		// Renewlet 自导出旧记录可能还没有 extra.import；当前用户内用原订阅 id 做二级匹配，保证导出再导入能 replace/skip。
		result.ByKey[importKeyString(importKey{Source: "renewlet", SourceID: row.Id})] = row
		extra := map[string]interface{}{}
		data, err := jsonBytesFromValue(row.Get("extra"))
		if err != nil || len(strings.TrimSpace(string(data))) == 0 {
			continue
		}
		if err := json.Unmarshal(data, &extra); err != nil {
			continue
		}
		key, err := importKeyFromExtra(extra)
		if err != nil {
			continue
		}
		result.ByKey[importKeyString(key)] = row
		if isLowConfidenceWallosKey(key) {
			result.AddLowConfidence(row)
		}
	}
	return result
}

func (matches importExistingMatches) AddLowConfidence(row *core.Record) {
	nameKey := lowConfidenceImportName(row.GetString("name"))
	if nameKey == "" {
		return
	}
	if matches.LowConfidenceDuplicates[nameKey] {
		return
	}
	if matches.LowConfidenceByName[nameKey] != nil {
		// 同名历史订阅一多，名称兜底就失去唯一性；后续必须走用户手动选择。
		delete(matches.LowConfidenceByName, nameKey)
		matches.LowConfidenceDuplicates[nameKey] = true
		return
	}
	matches.LowConfidenceByName[nameKey] = row
}

func (matches importExistingMatches) Resolve(key importKey, subscription importSubscription) (*core.Record, bool) {
	if existing := matches.ByKey[importKeyString(key)]; existing != nil {
		return existing, false
	}
	if !isLowConfidenceWallosKey(key) {
		return nil, false
	}
	nameKey := lowConfidenceImportName(subscription.Name)
	if nameKey == "" || matches.LowConfidenceDuplicates[nameKey] {
		return nil, false
	}
	return matches.LowConfidenceByName[nameKey], matches.LowConfidenceByName[nameKey] != nil
}

func importKeyFromExtra(extra map[string]interface{}) (importKey, error) {
	raw, ok := extra["import"].(map[string]interface{})
	if !ok {
		return importKey{}, errors.New("IMPORT_KEY_REQUIRED")
	}
	source, _ := raw["source"].(string)
	sourceID, _ := raw["sourceId"].(string)
	confidence, _ := raw["confidence"].(string)
	source = strings.TrimSpace(source)
	sourceID = strings.TrimSpace(sourceID)
	if source != "renewlet" && source != "wallos" {
		return importKey{}, errors.New("IMPORT_SOURCE_INVALID")
	}
	if sourceID == "" || len([]rune(sourceID)) > 256 {
		return importKey{}, errors.New("IMPORT_SOURCE_ID_INVALID")
	}
	if confidence != "" && confidence != "high" && confidence != "low" {
		return importKey{}, errors.New("IMPORT_CONFIDENCE_INVALID")
	}
	return importKey{Source: source, SourceID: sourceID, Confidence: confidence}, nil
}

func importSkippedIndexSet(indexes []int, subscriptionCount int) (map[int]bool, error) {
	result := map[int]bool{}
	for _, index := range indexes {
		if index < 0 || index >= subscriptionCount {
			return nil, errors.New("IMPORT_SKIP_INDEX_INVALID")
		}
		result[index] = true
	}
	return result, nil
}

func isLowConfidenceWallosKey(key importKey) bool {
	return key.Source == "wallos" && (key.Confidence == "low" || strings.HasPrefix(key.SourceID, "display:"))
}

func lowConfidenceImportName(value string) string {
	return strings.ToLower(strings.Join(strings.Fields(value), " "))
}

func importKeyString(key importKey) string {
	return key.Source + ":" + key.SourceID
}

func summarizeImportItems(items []importPreviewItem) importSummary {
	var summary importSummary
	for _, item := range items {
		summary.Total++
		switch item.Action {
		case "create":
			summary.Creates++
		case "replace":
			summary.Replaces++
		case "skip":
			summary.Skips++
		case "error":
			summary.Errors++
		}
		summary.Warnings += len(item.Warnings)
		if len(item.Errors) > 0 && item.Action != "error" {
			summary.Errors++
		}
	}
	return summary
}

func optionalString(value *string) string {
	if value == nil {
		return ""
	}
	return strings.TrimSpace(*value)
}

func isValidBillingCycle(value string) bool {
	switch value {
	case "weekly", "monthly", "quarterly", "semi-annual", "annual", "custom", "one-time":
		return true
	default:
		return false
	}
}

func isValidSubscriptionStatus(value string) bool {
	switch value {
	case "trial", "active", "expired", "paused", "cancelled":
		return true
	default:
		return false
	}
}

File: packages/cloudflare/src/import-export.ts
import {
  importApplyRequestSchema,
  importApplyResponseSchema,
  IMPORT_APPLY_SUBSCRIPTION_LIMIT,
  importPreviewRequestSchema,
  importPreviewResponseSchema,
  type ImportConflictMode,
  type ImportPayload,
  type ImportPreviewItem,
  type ImportSummary,
  type ImportSubscription,
} from "@renewlet/shared/schemas/import-export";
import { appSettingsSchema } from "@renewlet/shared/schemas/settings";
import { customConfigSchema } from "@renewlet/shared/schemas/custom-config";
import { cleanBuiltInIconSourceSettingsPatch, mergeBuiltInIconSourceSettings } from "@renewlet/shared/built-in-icons";
import { getSettings, listSubscriptions, newId, nowIso, parseJsonObject } from "./db";
import { requestLocale, json, readJsonWithLimit, HttpError, type AppLocale } from "./http";
import { serverText } from "./server-i18n";
import { requireAuth } from "./auth";
import { subscriptionRowValues, toSubscriptionRow } from "./subscriptions";
import type { Env, SubscriptionRow } from "./types";

const INSERT_SUBSCRIPTION_SQL = `
  INSERT INTO subscriptions (
    id, user_id, name, logo, price, currency, billing_cycle, custom_days, category, status, pinned, payment_method,
    start_date, next_billing_date, auto_calculate_next_billing_date, trial_end_date, website, notes, tags_json,
    reminder_days, repeat_reminder_enabled, repeat_reminder_interval, repeat_reminder_window, extra_json, created_at, updated_at
  ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
`;

const UPDATE_SUBSCRIPTION_SQL = `
  UPDATE subscriptions SET
    name = ?, logo = ?, price = ?, currency = ?, billing_cycle = ?, custom_days = ?, category = ?, status = ?,
    pinned = ?, payment_method = ?, start_date = ?, next_billing_date = ?, auto_calculate_next_billing_date = ?,
    trial_end_date = ?, website = ?, notes = ?, tags_json = ?, reminder_days = ?, repeat_reminder_enabled = ?,
    repeat_reminder_interval = ?, repeat_reminder_window = ?, extra_json = ?, updated_at = ?
  WHERE user_id = ? AND id = ?
`;

const IMPORT_JSON_LIMIT_BYTES = 50 * 1024 * 1024;
const IMPORT_WARNING_LOW_CONFIDENCE_KEY = "IMPORT_WARNING_LOW_CONFIDENCE_KEY";
const IMPORT_WARNING_LOW_CONFIDENCE_NAME_MATCHED = "IMPORT_WARNING_LOW_CONFIDENCE_NAME_MATCHED";

/** 导入预览只做当前用户范围内的冲突判断，不写 D1。 */
export async function previewImport(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requireAuth(request, env);
  const body = await readJsonWithLimit(request, importPreviewRequestSchema, locale, IMPORT_JSON_LIMIT_BYTES);
  assertValidSkipIndexes(body.skipIndexes, body.payload.subscriptions.length, locale);
  const existing = await listSubscriptions(env, auth.user.id);
  return json(importPreviewResponseSchema.parse(buildPreview(body.payload, body.conflictMode, existing, body.skipIndexes)));
}

/** 应用导入会重新计算 preview，避免客户端篡改 action 结果后直接写库。 */
export async function applyImport(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  const auth = await requireAuth(request, env);
  // apply 必须重新解析和预览请求体，不能信任浏览器上一轮 preview 的 action 结果。
  const body = await readJsonWithLimit(request, importApplyRequestSchema, locale, IMPORT_JSON_LIMIT_BYTES);
  assertApplyPayloadSize(body.payload.subscriptions.length, locale);
  // 导入只在当前登录用户范围内查重；payload 里的来源用户仅用于 extra.import 幂等键，不能变成 owner。
  assertValidSkipIndexes(body.skipIndexes, body.payload.subscriptions.length, locale);
  const existing = await listSubscriptions(env, auth.user.id);
  const preview = buildPreview(body.payload, body.conflictMode, existing, body.skipIndexes);
  if (preview.summary.errors > 0) {
    throw new HttpError(400, serverText(locale, "import.previewFailed"), "IMPORT_PREVIEW_FAILED", preview);
  }

  const timestamp = nowIso();
  const statements: D1PreparedStatement[] = [];
  const existingMatches = buildExistingImportMatches(existing);
  for (const item of preview.items) {
    if (item.action !== "create" && item.action !== "replace") continue;
    const source = body.payload.subscriptions[item.index];
    if (!source) continue;
    const { row: existingRow } = resolveExistingImportMatch(existingMatches, source.extra.import, source);
    // toSubscriptionRow 只接收订阅 allowlist 字段，user_id 始终来自当前 session，避免导入 payload 做 mass assignment。
    const row = toSubscriptionRow(
      existingRow?.id ?? newId("sub"),
      auth.user.id,
      source,
      existingRow?.created_at ?? timestamp,
      timestamp,
    );
    if (existingRow) {
      statements.push(env.DB.prepare(UPDATE_SUBSCRIPTION_SQL).bind(
        row.name,
        row.logo,
        row.price,
        row.currency,
        row.billing_cycle,
        row.custom_days,
        row.category,
        row.status,
        row.pinned,
        row.payment_method,
        row.start_date,
        row.next_billing_date,
        row.auto_calculate_next_billing_date,
        row.trial_end_date,
        row.website,
        row.notes,
        row.tags_json,
        row.reminder_days,
        row.repeat_reminder_enabled,
        row.repeat_reminder_interval,
        row.repeat_reminder_window,
        row.extra_json,
        timestamp,
        auth.user.id,
        existingRow.id,
      ));
    } else {
      statements.push(env.DB.prepare(INSERT_SUBSCRIPTION_SQL).bind(...subscriptionRowValues(row)));
    }
  }

  if (body.payload.settings) {
    const current = await getSettings(env, auth.user.id);
    const next = appSettingsSchema.parse({
      ...current,
      ...body.payload.settings,
      builtInIconSources: mergeBuiltInIconSourceSettings(current.builtInIconSources, cleanBuiltInIconSourceSettingsPatch(body.payload.settings.builtInIconSources)),
    });
    statements.push(env.DB.prepare(`
      INSERT INTO settings (user_id, settings_json, created_at, updated_at)
      VALUES (?, ?, ?, ?)
      ON CONFLICT(user_id) DO UPDATE SET settings_json = excluded.settings_json, updated_at = excluded.updated_at
    `).bind(auth.user.id, JSON.stringify(next), timestamp, timestamp));
  }
  if (body.payload.customConfig) {
    const nextConfig = customConfigSchema.parse(body.payload.customConfig);
    statements.push(env.DB.prepare(`
      INSERT INTO custom_configs (user_id, config_json, created_at, updated_at)
      VALUES (?, ?, ?, ?)
      ON CONFLICT(user_id) DO UPDATE SET config_json = excluded.config_json, updated_at = excluded.updated_at
    `).bind(auth.user.id, JSON.stringify(nextConfig), timestamp, timestamp));
  }

  if (statements.length > 0) {
    // D1 batch 在同一事务里执行；导入要么整体写入，要么让调用方看到明确失败。
    await env.DB.batch(statements);
  }
  return json(importApplyResponseSchema.parse({ ok: true, ...preview }));
}

function assertApplyPayloadSize(count: number, locale: AppLocale): void {
  if (count > IMPORT_APPLY_SUBSCRIPTION_LIMIT) {
    throw new HttpError(400, serverText(locale, "import.invalid"), "IMPORT_TOO_MANY_SUBSCRIPTIONS");
  }
}

type PreviewResult = {
  summary: ImportSummary;
  items: ImportPreviewItem[];
  includesSettings: boolean;
  includesCustomConfig: boolean;
};

function buildPreview(payload: ImportPayload, conflictMode: ImportConflictMode, existing: SubscriptionRow[], skipIndexes: number[]): PreviewResult {
  const existingMatches = buildExistingImportMatches(existing);
  const skippedIndexes = new Set(skipIndexes);
  const seenPayloadKeys = new Set<string>();
  const items = payload.subscriptions.map((subscription, index) => {
    const importKey = subscription.extra.import;
    const warnings: string[] = [];
    const errors: string[] = [];
    if (importKey.confidence === "low") {
      warnings.push(IMPORT_WARNING_LOW_CONFIDENCE_KEY);
    }
    if (skippedIndexes.has(index)) {
      return {
        index,
        name: subscription.name,
        source: importKey.source,
        sourceId: importKey.sourceId,
        action: "skip",
        warnings,
        errors,
      } satisfies ImportPreviewItem;
    }
    const keyString = importKeyString(importKey);
    if (seenPayloadKeys.has(keyString)) {
      // 同一个导入文件内部重复 sourceId 必须先失败；否则 replace 会把两条 payload 写到同一订阅上。
      errors.push("IMPORT_SOURCE_ID_DUPLICATE");
    }
    seenPayloadKeys.add(keyString);
    const { row: existingRow, fallback } = resolveExistingImportMatch(existingMatches, importKey, subscription);
    if (fallback) {
      // Wallos display:* 是低置信桥接，只给用户 warning；真正写入仍保留原 import key 方便后续精确替换。
      warnings.push(IMPORT_WARNING_LOW_CONFIDENCE_NAME_MATCHED);
    }
    const action = errors.length > 0 ? "error" : existingRow ? (conflictMode === "replace" ? "replace" : "skip") : "create";
    return {
      index,
      name: subscription.name,
      source: importKey.source,
      sourceId: importKey.sourceId,
      ...(existingRow ? { existingId: existingRow.id } : {}),
      action,
      warnings,
      errors,
    } satisfies ImportPreviewItem;
  });
  return {
    summary: summarize(items),
    items,
    includesSettings: Boolean(payload.settings),
    includesCustomConfig: Boolean(payload.customConfig),
  };
}

function summarize(items: ImportPreviewItem[]): ImportSummary {
  return items.reduce<ImportSummary>((summary, item) => ({
    total: summary.total + 1,
    creates: summary.creates + (item.action === "create" ? 1 : 0),
    replaces: summary.replaces + (item.action === "replace" ? 1 : 0),
    skips: summary.skips + (item.action === "skip" ? 1 : 0),
    errors: summary.errors + (item.errors.length > 0 ? 1 : 0),
    warnings: summary.warnings + item.warnings.length,
  }), { total: 0, creates: 0, replaces: 0, skips: 0, errors: 0, warnings: 0 });
}

type ExistingImportMatches = {
  byKey: Map<string, SubscriptionRow>;
  lowConfidenceByName: Map<string, SubscriptionRow | null>;
};

function buildExistingImportMatches(rows: SubscriptionRow[]): ExistingImportMatches {
  const result: ExistingImportMatches = {
    byKey: new Map<string, SubscriptionRow>(),
    lowConfidenceByName: new Map<string, SubscriptionRow | null>(),
  };
  for (const row of rows) {
    // Renewlet 自导入要兼容“已有订阅还没有 extra.import”的当前数据；只在当前用户查询结果内按真实 id 建二级键。
    result.byKey.set(`renewlet:${row.id}`, row);
    const extra = parseJsonObject(row.extra_json);
    const importValue = extra["import"];
    if (!isImportKey(importValue)) continue;
    result.byKey.set(importKeyString(importValue), row);
    if (isLowConfidenceWallosKey(importValue)) {
      addLowConfidenceExisting(result, row);
    }
  }
  return result;
}

function addLowConfidenceExisting(matches: ExistingImportMatches, row: SubscriptionRow): void {
  const nameKey = lowConfidenceImportName(row.name);
  if (!nameKey) return;
  if (matches.lowConfidenceByName.has(nameKey)) {
    // 同名历史订阅不唯一时禁用名称兜底，宁可让用户手动处理，也不要误 replace。
    matches.lowConfidenceByName.set(nameKey, null);
    return;
  }
  matches.lowConfidenceByName.set(nameKey, row);
}

function resolveExistingImportMatch(
  matches: ExistingImportMatches,
  key: ImportSubscription["extra"]["import"],
  subscription: ImportSubscription,
): { row: SubscriptionRow | undefined; fallback: boolean } {
  const exact = matches.byKey.get(importKeyString(key));
  if (exact) return { row: exact, fallback: false };
  if (!isLowConfidenceWallosKey(key)) return { row: undefined, fallback: false };
  const fallback = matches.lowConfidenceByName.get(lowConfidenceImportName(subscription.name));
  return { row: fallback ?? undefined, fallback: Boolean(fallback) };
}

function isImportKey(value: unknown): value is ImportSubscription["extra"]["import"] {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const record = value as Record<string, unknown>;
  return (record["source"] === "renewlet" || record["source"] === "wallos")
    && typeof record["sourceId"] === "string"
    && (record["confidence"] === undefined || record["confidence"] === "high" || record["confidence"] === "low");
}

function assertValidSkipIndexes(indexes: number[], subscriptionCount: number, locale: ReturnType<typeof requestLocale>): void {
  if (indexes.some((index) => index < 0 || index >= subscriptionCount)) {
    throw new HttpError(400, serverText(locale, "import.skipIndexInvalid"), "IMPORT_SKIP_INDEX_INVALID");
  }
}

function isLowConfidenceWallosKey(value: ImportSubscription["extra"]["import"]): boolean {
  return value.source === "wallos" && (value.confidence === "low" || value.sourceId.startsWith("display:"));
}

function lowConfidenceImportName(value: string): string {
  return value.trim().toLowerCase().replace(/\s+/g, " ");
}

function importKeyString(value: ImportSubscription["extra"]["import"]): string {
  return `${value.source}:${value.sourceId}`;
}
